fix(oxc_codegen): don't emit empty span mapping

hi-ogawa opened the pull request:
Currently empty span `SPAN` is generated and this is likely problematic as it gives false mapping for original source's first line.

Here is how it affects rolldown sourcemap snapshots:
- https://github.com/rolldown/rolldown/pull/2888

---

Interestingly, esbuild leaves mapping for module wrapper code (e.g. `var require_xxx = __commonJS(...)`), but not sure if it's designed so: [esbuild.try](https://esbuild.github.io/try/#YgAwLjI0LjAALS1mb3JtYXQ9ZXNtIC0tYnVuZGxlIC0tc291cmNlbWFwAGUAZW50cnkuanMALy8geApyZXF1aXJlKCcuL2Nqcy5qcycpOwovLyB4CnZhciBjanMgPSByZXF1aXJlKCcuL2Nqcy5qcycpOwAAY2pzLmpzAC8vIHgKLy8geAptb2R1bGUuZXhwb3J0cyA9IHt9), [sourcemap viz](https://evanw.github.io/source-map-visualization/#Nzk5AHZhciBfX2dldE93blByb3BOYW1lcyA9IE9iamVjdC5nZXRPd25Qcm9wZXJ0eU5hbWVzOwp2YXIgX19jb21tb25KUyA9IChjYiwgbW9kKSA9PiBmdW5jdGlvbiBfX3JlcXVpcmUoKSB7CiAgcmV0dXJuIG1vZCB8fCAoMCwgY2JbX19nZXRPd25Qcm9wTmFtZXMoY2IpWzBdXSkoKG1vZCA9IHsgZXhwb3J0czoge30gfSkuZXhwb3J0cywgbW9kKSwgbW9kLmV4cG9ydHM7Cn07CgovLyBjanMuanMKdmFyIHJlcXVpcmVfY2pzID0gX19jb21tb25KUyh7CiAgImNqcy5qcyIoZXhwb3J0cywgbW9kdWxlKSB7CiAgICBtb2R1bGUuZXhwb3J0cyA9IHt9OwogIH0KfSk7CgovLyBlbnRyeS5qcwpyZXF1aXJlX2NqcygpOwp2YXIgY2pzID0gcmVxdWlyZV9janMoKTsKLy8jIHNvdXJjZU1hcHBpbmdVUkw9ZGF0YTphcHBsaWNhdGlvbi9qc29uO2Jhc2U2NCxld29nSUNKMlpYSnphVzl1SWpvZ015d0tJQ0FpYzI5MWNtTmxjeUk2SUZzaVkycHpMbXB6SWl3Z0ltVnVkSEo1TG1weklsMHNDaUFnSW5OdmRYSmpaWE5EYjI1MFpXNTBJam9nV3lJdkx5QjRYRzR2THlCNFhHNXRiMlIxYkdVdVpYaHdiM0owY3lBOUlIdDlJaXdnSWk4dklIaGNibkpsY1hWcGNtVW9KeTR2WTJwekxtcHpKeWs3WEc0dkx5QjRYRzUyWVhJZ1kycHpJRDBnY21WeGRXbHlaU2duTGk5amFuTXVhbk1uS1RzaVhTd0tJQ0FpYldGd2NHbHVaM01pT2lBaU96czdPenM3UVVGQlFUdEJRVUZCTzBGQlJVRXNWMEZCVHl4VlFVRlZMRU5CUVVNN1FVRkJRVHRCUVVGQk96czdRVU5FYkVJN1FVRkZRU3hKUVVGSkxFMUJRVTA3SWl3S0lDQWlibUZ0WlhNaU9pQmJYUXA5Q2c9PQoyNTkAeyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImNqcy5qcyIsImVudHJ5LmpzIl0sInNvdXJjZXNDb250ZW50IjpbIi8vIHhcbi8vIHhcbm1vZHVsZS5leHBvcnRzID0ge30iLCIvLyB4XG5yZXF1aXJlKCcuL2Nqcy5qcycpO1xuLy8geFxudmFyIGNqcyA9IHJlcXVpcmUoJy4vY2pzLmpzJyk7Il0sIm1hcHBpbmdzIjoiOzs7Ozs7QUFBQTtBQUFBO0FBRUEsV0FBTyxVQUFVLENBQUM7QUFBQTtBQUFBOzs7QUNEbEI7QUFFQSxJQUFJLE1BQU07IiwibmFtZXMiOltdfQ==)

hi-ogawa committed:
fix(oxc_codegen): don't emit empty span mapping

```
ast-grep -p '$A.add_source_mapping($B.start)' -r '$A.add_source_mapping_by_span($B, $B.start)' -U
```

hi-ogawa committed:
refactor: replace more

Boshen commented on crates/oxc_codegen/src/lib.rs:
```diff
@@ -639,13 +639,19 @@ impl<'a> Codegen<'a> {
     }
 
-    fn add_source_mapping(&mut self, position: u32) {
+    fn add_source_mapping(&mut self, span: Span, position: u32) {
```
Can't you use `span.start` as `position` if all positions are `span.start`?

hi-ogawa replied:
There are a few cases `span.end` is explicitly emitted, for example:

https://github.com/oxc-project/oxc/blob/41d401148b773bd7256f74e7cceb95fe5576440f/crates/oxc_codegen/src/gen.rs#L815

In total, there are about 6 cases (span.end) vs. 90 cases (span.start), so I can probably add `add_source_mapping_end` dedicated for those minority. Would it be preferred?

Boshen replied:
Yes, `add_source_mapping_end` seems better.

hi-ogawa committed:
wip: add_source_mapping_end

```
ast-grep -p '$A.add_source_mapping($B, $C.end)' -r '$A.add_source_mapping_end($B)' -U
```

hi-ogawa committed:
wip: print_block_end

hi-ogawa committed:
wip: strip start

```
ast-grep -p '$A.add_source_mapping($B, $C)' -r '$A.add_source_mapping($B)' -U
```

hi-ogawa committed:
wip: fixup
